core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -558,10 +558,11 @@ class KafkaController(val config: KafkaConfig, zkClient: KafkaZkClient, time: Ti
    * @throws IllegalStateException if a partition is not in `partitionsBeingReassigned`
    */
   private def maybeTriggerPartitionReassignment(topicPartitions: Set[TopicPartition]) {
+    val partitionsToBeRemovedFromReassignment = scala.collection.mutable.Set.empty[TopicPartition]
     topicPartitions.foreach { tp =>
       if (topicDeletionManager.isTopicQueuedUpForDeletion(tp.topic)) {
         error(s"Skipping reassignment of $tp since the topic is currently being deleted")
-        removePartitionFromReassignedPartitions(tp)
+        partitionsToBeRemovedFromReassignment.add(tp)
       } else {
         val reassignedPartitionContext = controllerContext.partitionsBeingReassigned.get(tp).getOrElse {
           throw new IllegalStateException(s"Initiating reassign replicas for partition $tp not present in " +
```

Comment:
What if IllegalStateException thrown here?   we may miss calling removePartitionsFromReassignedPartitions. We may need to use try..finally here?

Reply:
Right. If IllegalStateException is thrown, we will not call removePartitionsFromReassignedPartitions. This is the same behavior as before. IllegalStateException indicates a bug and we generally don't expect to properly recover from such state.